From 81a1d9bcae0e3dae7d1e89f4e4fa62e9a230b12d Mon Sep 17 00:00:00 2001
From: Pranav Vashi <neobuddy89@gmail.com>
Date: Sat, 17 Jan 2026 20:33:02 +0530
Subject: [PATCH 06/30] Launcher3: Add a toggle to disable hotseat searchbar

* @jhonboy121:
  * cache function return values from Utilities class instead of calling over and over.
  * show a toast when switch is toggled, using the same string for feature flags but made translatable.
  * set proper checked value on preference init.

Change-Id: I93d973f9c4c04d02608f5f657837397959430a43
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
Signed-off-by: jhonboy121 <alfredmathew05@gmail.com>
---
 res/values/extra_strings.xml                  |  3 ++
 res/xml/launcher_preferences.xml              |  7 ++++
 src/com/android/launcher3/Hotseat.java        |  7 ++--
 src/com/android/launcher3/Utilities.java      | 23 ++++++++++++
 .../launcher3/settings/SettingsActivity.java  | 35 +++++++++++++++++++
 5 files changed, 73 insertions(+), 2 deletions(-)

diff --git a/res/values/extra_strings.xml b/res/values/extra_strings.xml
index 0800bc04c2..f92e4cb791 100644
--- a/res/values/extra_strings.xml
+++ b/res/values/extra_strings.xml
@@ -4,4 +4,7 @@
     <string name="label_search">Search</string>
     <string name="label_lens">Google Lens</string>
     <string name="label_voice_search">Voice search</string>
+    <string name="dock_search_title">Google search bar</string>
+    <string name="dock_search_summary">Search bar in the bottom dock</string>
+    <string name="flag_wont_be_applied">Flag won\'t be applied until you restart launcher</string>
 </resources>
diff --git a/res/xml/launcher_preferences.xml b/res/xml/launcher_preferences.xml
index a2876d9dc1..778556315b 100644
--- a/res/xml/launcher_preferences.xml
+++ b/res/xml/launcher_preferences.xml
@@ -103,4 +103,11 @@
         android:defaultValue="true"
         android:persistent="true" />
 
+    <SwitchPreferenceCompat
+        android:key="pref_dock_search"
+        android:title="@string/dock_search_title"
+        android:summary="@string/dock_search_summary"
+        android:defaultValue="true"
+        launcher:iconSpaceReserved="false" />
+
 </androidx.preference.PreferenceScreen>
diff --git a/src/com/android/launcher3/Hotseat.java b/src/com/android/launcher3/Hotseat.java
index 239e8afb22..755d2a996b 100644
--- a/src/com/android/launcher3/Hotseat.java
+++ b/src/com/android/launcher3/Hotseat.java
@@ -109,8 +109,11 @@ public class Hotseat extends CellLayout implements Insettable {
             mQsb = LayoutInflater.from(context).inflate(R.layout.qsb_container_hotseat, this,
                     false);
         } else {
-            mQsb = LayoutInflater.from(context).inflate(R.layout.search_container_hotseat, this,
-                    false);
+            if (Utilities.showQSB(context)) {
+                mQsb = LayoutInflater.from(context).inflate(R.layout.search_container_hotseat, this, false);
+            } else {
+                mQsb = LayoutInflater.from(context).inflate(R.layout.empty_view, this, false);
+        }
         }
 
         addView(mQsb);
diff --git a/src/com/android/launcher3/Utilities.java b/src/com/android/launcher3/Utilities.java
index 2dd683c041..cc330e676a 100644
--- a/src/com/android/launcher3/Utilities.java
+++ b/src/com/android/launcher3/Utilities.java
@@ -36,9 +36,11 @@ import android.app.WallpaperManager;
 import android.content.Context;
 import android.content.pm.LauncherActivityInfo;
 import android.content.pm.LauncherApps;
+import android.content.pm.PackageManager;
 import android.content.pm.ShortcutInfo;
 import android.content.res.Configuration;
 import android.content.res.Resources;
+import android.content.SharedPreferences;
 import android.graphics.Color;
 import android.graphics.ColorFilter;
 import android.graphics.LightingColorFilter;
@@ -161,6 +163,9 @@ public final class Utilities {
     @IntDef({TRANSLATE_UP, TRANSLATE_DOWN, TRANSLATE_LEFT, TRANSLATE_RIGHT})
     public @interface AdjustmentDirection{}
 
+    public static final String GSA_PACKAGE = "com.google.android.googlequicksearchbox";
+    public static final String KEY_DOCK_SEARCH = "pref_dock_search";
+
     /**
      * Returns true if theme is dark.
      */
@@ -1065,4 +1070,22 @@ public final class Utilities {
     public static boolean isWorkspaceEditAllowed(Context context) {
         return !LauncherPrefs.WORKSPACE_LOCK.get(context);
     }
+
+    public static boolean isGSAEnabled(Context context) {
+        try {
+            return context.getPackageManager().getApplicationInfo(GSA_PACKAGE, 0).enabled;
+        } catch (PackageManager.NameNotFoundException e) {
+            return false;
+        }
+    }
+
+    public static boolean showQSB(Context context) {
+        return isGSAEnabled(context) && isQSBEnabled(context);
+    }
+
+    private static boolean isQSBEnabled(Context context) {
+        SharedPreferences prefs = LauncherPrefs.getPrefs(context.getApplicationContext());
+        return prefs.getBoolean(KEY_DOCK_SEARCH, true);
+    }
+
 }
diff --git a/src/com/android/launcher3/settings/SettingsActivity.java b/src/com/android/launcher3/settings/SettingsActivity.java
index c3e408d26c..767913270b 100644
--- a/src/com/android/launcher3/settings/SettingsActivity.java
+++ b/src/com/android/launcher3/settings/SettingsActivity.java
@@ -28,9 +28,11 @@ import static com.android.launcher3.InvariantDeviceProfile.TYPE_TABLET;
 import static com.android.launcher3.states.RotationHelper.ALLOW_ROTATION_PREFERENCE_KEY;
 
 import android.app.Activity;
+import android.content.Context;
 import android.content.Intent;
 import android.content.pm.ActivityInfo;
 import android.content.pm.LauncherApps;
+import android.content.pm.PackageManager;
 import android.content.res.Resources;
 import android.net.Uri;
 import android.os.Bundle;
@@ -38,6 +40,7 @@ import android.provider.Settings;
 import android.text.TextUtils;
 import android.view.MenuItem;
 import android.view.View;
+import android.widget.Toast;
 
 import androidx.annotation.Nullable;
 import androidx.annotation.VisibleForTesting;
@@ -53,6 +56,7 @@ import androidx.preference.PreferenceFragmentCompat.OnPreferenceStartScreenCallb
 import androidx.preference.PreferenceGroup;
 import androidx.preference.PreferenceGroup.PreferencePositionCallback;
 import androidx.preference.PreferenceScreen;
+import androidx.preference.SwitchPreferenceCompat;
 import androidx.recyclerview.widget.RecyclerView;
 
 import com.android.launcher3.BuildConfig;
@@ -60,11 +64,14 @@ import com.android.launcher3.Flags;
 import com.android.launcher3.InvariantDeviceProfile;
 import com.android.launcher3.LauncherFiles;
 import com.android.launcher3.R;
+import com.android.launcher3.config.FeatureFlags;
 import com.android.launcher3.lineage.LineageUtils;
 import com.android.launcher3.lineage.trust.TrustAppsActivity;
 import com.android.launcher3.states.RotationHelper;
 import com.android.launcher3.util.DisplayController;
+import com.android.launcher3.util.PluginManagerWrapper;
 import com.android.launcher3.util.SettingsCache;
+import com.android.launcher3.Utilities;
 import com.android.settingslib.widget.SettingsBasePreferenceFragment;
 import com.android.settingslib.widget.SettingsThemeHelper;
 
@@ -201,6 +208,10 @@ public class SettingsActivity extends FragmentActivity
 
         private boolean mPreferenceHighlighted = false;
 
+        protected static final String GSA_PACKAGE = "com.google.android.googlequicksearchbox";
+
+        private Preference mShowGoogleAppPref, mDockSearchPref;
+
         @Override
         public void onCreate(@Nullable Bundle savedInstanceState) {
             if (BuildConfig.IS_DEBUG_DEVICE) {
@@ -376,10 +387,33 @@ public class SettingsActivity extends FragmentActivity
                 case KEY_SUGGESTIONS:
                     return launcherApps != null &&
                             launcherApps.isPackageEnabled(SUGGESTIONS_PACKAGE, myUserHandle());
+                case Utilities.KEY_DOCK_SEARCH:
+                    mDockSearchPref = preference;
+                    ((SwitchPreferenceCompat) mDockSearchPref).setChecked(Utilities.showQSB(getContext()));
+                    mDockSearchPref.setOnPreferenceChangeListener((pref, newValue) -> {
+                        Toast.makeText(getContext(), R.string.flag_wont_be_applied, Toast.LENGTH_LONG).show();
+                        return true;
+                    });
+                    updateIsGoogleAppEnabled();
+                    return true;
             }
             return true;
         }
 
+        public static boolean isGSAEnabled(Context context) {
+            try {
+                return context.getPackageManager().getApplicationInfo(GSA_PACKAGE, 0).enabled;
+            } catch (PackageManager.NameNotFoundException e) {
+                return false;
+            }
+        }
+
+        private void updateIsGoogleAppEnabled() {
+            if (mShowGoogleAppPref != null) {
+                mShowGoogleAppPref.setEnabled(isGSAEnabled(getContext()));
+            }
+        }
+
         @Override
         public void onResume() {
             super.onResume();
@@ -395,6 +429,7 @@ public class SettingsActivity extends FragmentActivity
             if (mRestartOnResume) {
                 recreateActivityNow();
             }
+            updateIsGoogleAppEnabled();
         }
 
         @Override
-- 
2.52.0

